### ISSUE TITLE
Not support mIRC background color?

### ISSUE THREAD

**Issue by Mosesofmason** [MCMrARM/revolution-irc#329]
The mIRC background color does not display correctly. 

e.g.
<CODE><COLOR>,<COLOR>

CODE is ASCII 0x03
COLOR is 00~99 IRC color characters
The first <COLOR> is the foreground, and the second <COLOR> is the background color (if sent).

**Comment by kurahaupo** [comment]
This seems odd to say the least. (I personally use revolution IRC every day, and it displays foreground and background colours just fine at least 99% of the time. I know there are occasional glitches, but they're almost always when sending, not receiving.)

* Does it affect sent messages, received messages, or both?
* Does this happen always or just occasionally?
* Does this happen in all channels or just some? Does it affect private messages?
* Which colour codes have you tested? All eight? Any that are neither black nor white?

(Without testing multiple colours, it's difficult to know whether the lack of change is simply because the selected colour is already the same as the default background. The colour codes for "black" and "white" are ambiguous because clients differ which way round they go and whether they change if you invert the colour scheme.)

Are you sure you're not in a channel that has mIRC attribute filtering turned on? (Sometimes the channel owners simply don't like coloured text.)

Can you give an example of a byte stream that does not display correctly, as recorded raw logs if possible?
Can you provide a screenshot?

**Comment by Mosesofmason** [comment]
@kurahaupo Yes, it happens always in all channels. I just tested `0x03` `58` `,` `66` `<some random test words>`, below is the screenshots comparison between mIRC and  Revolution IRC

mIRC:
![image](https://user-images.githubusercontent.com/161946/192929780-c0bebc9c-79e6-401d-aea2-9b58b2a65d3f.png)

Revolution IRC:
![image](https://user-images.githubusercontent.com/161946/192930396-cc6b6822-875c-4342-83ca-a1e3971de82a.png)

**Comment by kurahaupo** [comment]
It appears that you're using a non-standard extension, as only colours for codes `00` through `15` are specified in any mIRC documentation I could find.

Revolution IRC takes a conservative approach to security, and - as a security measure - where a message contains any invalid attribute code, no interpretation of _any_ attribute is performed on the displayed text. Accordingly the numeric codes `16` through `98` are treated as undefined attribute codes. This is enforced by a [check for invalid colour codes](/MCMrARM/revolution-irc/blob/master/app/src/main/java/io/mrarm/irc/util/IRCColorUtils.java#L185).

I will review this policy with regard to colour codes, with a view to treating any undefined colours as "default" (equivalent to `99`).

On the one hand the [mirc.com website](https://www.mirc.com/help/html/index.html?control_codes.html) claims that codes `16` through `98` apply to _"all IRC users"_ rather than _"all **m**IRC users"_, but on the other hand, it fails to document the RGB values for these codes, making it difficult for other IRC clients to provide this supposedly universal behaviour.

If you can cite a reliable source for a table of RGB values for the colour codes 16 through 98 I will be more than happy to add them to Revolution IRC.

**Comment by kurahaupo** [comment]
Thanks for the table and the link to the forum. I will see about creating a pull request to add them shortly.

There's one error in the proposal: it refers to `99` as _transparent_. That certainly doesn't match the meaning ascribed to "transparent" in the CSS standard, nor in ordinary English, since the effect of `99` is different for the foreground and background colours. It would be more appropriate to say that `99` reverts to a _default_ or _initial_ colour.

There's also an error in the description where it says _These colors are based on the VGA/XP/ANSI color table._ Whilst it has the same range of colours, that's down to purely physical constraints. It's clearly _not_ based on VGA or ANSI as the order of the table is quite different. (VGA & ANSI use bitwise chromatic mixing; the primary colours red, green & blue have codes 1, 2 & 4, "bright" has code 8, and the other colours have the appropriate bitwise sums of those numbers.)

Just for the sake of future readers, I note that the linked page references a _proposal for an extension_ back in 2012. Wider adoption didn't occur until around 2017, and standardization is seemingly still pending. (But if anyone can point me at an RFC or other formal standard for the mIRC extended colours, I would much appreciate it.)

All that said, the mIRC colour encoding has always been ambiguous and fragile, and in hindsight the community would have been far better served by formally adopting ANSI-style escape sequences as the preferred colour encoding, as they are a proven technology, don't have weird ambiguities that conflate attributes and text, are readily extensible, and were already accepted by clients well before mIRC came along, and still are accepted by some.

**Comment by Skpij** [comment]
Helo

**Comment by kurahaupo** [comment]
@Skpij I see that you're new to GitHub.

It's important to understand that this is not an interactive forum, and there will typically be days or even weeks between messages. You do not need to introduce yourself or "ask to be allowed to speak".

On the contrary, please skip any pleasantries and start directly by stating the problem you're having with RevolutionIRC (describe what you expect, and how it fails to do meet your expectations), and then any questions or suggestions.

If you don't have any interest in RevolutionIRC, and you're simply looking to learn about GitHub, then I suggest searching for software packages that you _are_ using.

**Comment by kurahaupo** [comment]
I've submitted PR #332 to add support for mIRC extended colour codes, plus ANSI colour codes, and ANSI attributes (bold, italic, underline, inverse)

**Comment by Mosesofmason** [comment]
Hi, when will this change live in the store (Google Play or F-Droid)?
